File: s3-accelerator/python/sse_kms_suite.py
# SPDX-License-Identifier: Apache-2.0
"""SSE-KMS on the Python path, held to the JVM path's rule (ADR-0023 D13.3).

The tier is SHARED between the two clients by design (the cross-language
drill asserts it), so a guarantee that differs by language is not a guarantee:
a Python reader caching KMS plaintext that the JVM reader refuses to cache
leaves exactly the hole the JVM rule was written to close.

Every "it did not cache" below is paired with proof that the same client DOES
cache a plain object, because a client that cached nothing at all would pass
every bypass assertion on its own.
"""

from __future__ import annotations

import hashlib
import sys

import redis as redis_lib

sys.path.insert(0, ".")
import flint_accel

OK = [True]


def check(cond, label):
    OK[0] &= bool(cond)
    print(f"[{'ok' if cond else 'FAIL'}] {label}")


def expect(key, off, n, gen=0):
    return bytes(hashlib.md5(f"{key}:{gen}:{(off + i) // 16}".encode()).digest()[(off + i) % 16]
                 for i in range(n))


def fs_for(endpoint, tier, cache_kms=False):
    return flint_accel.FlintS3FileSystem(
        skip_instance_cache=True, anon=False, key="p", secret="p",
        client_kwargs={"endpoint_url": endpoint, "region_name": "us-east-1"},
        tier_uri=tier, cache_sse_kms=cache_kms)


def read(f, key, off, n):
    with f.open(f"s3://bucket/{key}", "rb") as h:
        h.seek(off)
        return h.read(n)


def main():
    kms_ep = sys.argv[1] if len(sys.argv) > 1 else "http://127.0.0.1:9530"
    plain_ep = sys.argv[2] if len(sys.argv) > 2 else "http://127.0.0.1:9531"
    tier = sys.argv[3] if len(sys.argv) > 3 else "redis://127.0.0.1:9399"
    rc = redis_lib.Redis.from_url(tier)
    KEY, LEN = "data/000001.bin", 200_000

    # --- 1. default: a KMS object must not touch the tier --------------------
    rc.flushall()
    f = fs_for(kms_ep, tier)
    got = read(f, KEY, 0, LEN)
    check(got == expect(KEY, 0, LEN), "KMS object still reads CORRECTLY (bypass is not a failure)")
    n = len(rc.keys("*"))
    check(n == 0, f"and NOTHING reached the tier -- no chunks, no metadata ({n} keys)")
    c = f.counters
    check(c["kms_bypassed"] > 0,
          f"armed: it was the KMS path that bypassed ({c['kms_bypassed']}), not some other bypass")
    check(c["kms_undetectable"] == 0, "armed: detection actually worked -- 0 undetectable")

    # --- 2. the control that carries check 1 ---------------------------------
    rc.flushall()
    f2 = fs_for(plain_ep, tier)
    got2 = read(f2, KEY, 0, LEN)
    check(got2 == expect(KEY, 0, LEN), "control: a PLAIN object reads correctly")
    n2 = len(rc.keys("c1/*"))
    check(n2 > 0,
          f"control: and the same client DOES cache it ({n2} chunks) -- so check 1 "
          "measured the KMS rule, not a broken cache")
    check(f2.counters["kms_bypassed"] == 0, "control: and nothing was KMS-bypassed")

    # --- 3. opt-in -----------------------------------------------------------
    rc.flushall()
    f3 = fs_for(kms_ep, tier, cache_kms=True)
    got3 = read(f3, KEY, 0, LEN)
    check(got3 == expect(KEY, 0, LEN), "opt-in: the KMS object reads correctly")
    n3 = len(rc.keys("c1/*"))
    check(n3 > 0, f"OPT-IN WORKS: cache_sse_kms=True caches the KMS object ({n3} chunks)")
    check(f3.counters["kms_bypassed"] == 0, "armed: and nothing was bypassed once opted in")

    # --- 4. and the two clients agree on the KEY, not merely on the policy ---
    # The whole point of a shared tier is that the JVM client can read what
    # this one wrote. A Python-only bypass rule that used a different key
    # prefix would pass every check above and still be useless.
    sample = rc.keys("c1/*")[0].decode()
    check(sample.startswith("c1/") and sample.count("/") == 2,
          f"opt-in wrote the SHARED key shape the JVM client reads ({sample})")

    print("SSE-KMS PYTHON SUITE PASSED" if OK[0] else "SSE-KMS PYTHON SUITE FAILED")
    return 0 if OK[0] else 1


if __name__ == "__main__":
    sys.exit(main())

File: s3-accelerator/jvm-spike/src/main/java/ai/crestway/flintaccel/client/SseKmsSuite.java
// SPDX-License-Identifier: Apache-2.0
package ai.crestway.flintaccel.client;

import java.net.URI;
import java.util.Arrays;

import io.lettuce.core.RedisClient;
import io.lettuce.core.api.StatefulRedisConnection;
import io.lettuce.core.codec.ByteArrayCodec;

import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.s3.S3AsyncClient;

import software.amazon.s3.analyticsaccelerator.S3SdkObjectClient;

/**
 * SSE-KMS bypasses the cache unless the customer opts in (ADR-0023 D13.3).
 *
 * Unlike SSE-C, this is not a key-handling question -- S3 decrypts SSE-KMS
 * server-side and hands us plaintext lawfully. It is a question about two
 * properties the customer paid for and would silently lose:
 *
 *   the KMS grant stops being the access gate, because anyone who can read the
 *   tier reads the plaintext without holding kms:Decrypt; and
 *
 *   every cache hit is a decrypt that never reaches CloudTrail. For many
 *   customers that audit trail IS the compliance requirement, and no
 *   mitigation restores it -- it is a property of not calling KMS at all.
 *
 * Default off. Opt-in after the customer's own review. Never a surprise.
 *
 * The checks below are paired throughout: every "it did not cache" is
 * accompanied by proof that the same client DOES cache a non-KMS object, since
 * a client that cached nothing at all would pass every bypass assertion.
 */
public final class SseKmsSuite {

  static boolean ok = true;

  static void check(boolean c, String label) {
    ok &= c;
    System.out.printf("[%s] %s%n", c ? "ok" : "FAIL", label);
  }

  static S3AsyncClient sdk(String endpoint) {
    return S3AsyncClient.builder()
        .endpointOverride(URI.create(endpoint)).region(Region.US_EAST_1)
        .credentialsProvider(StaticCredentialsProvider.create(
            AwsBasicCredentials.create("k", "k")))
        .forcePathStyle(true).build();
  }

  public static void main(String[] args) throws Exception {
    // arg0 = a KMS-reporting origin, arg1 = a plain one, arg2 = tier
    String kmsEp   = args.length > 0 ? args[0] : "http://127.0.0.1:9530";
    String plainEp = args.length > 1 ? args[1] : "http://127.0.0.1:9531";
    String tierUri = args.length > 2 ? args[2] : "redis://127.0.0.1:9399";
    Suite.endpoint = kmsEp;

    RedisClient rc = RedisClient.create(tierUri);
    StatefulRedisConnection<byte[], byte[]> conn = rc.connect(new ByteArrayCodec());
    final String KEY = "data/000001.bin";
    final int LEN = 200_000;

    S3AsyncClient kmsS3 = sdk(kmsEp), plainS3 = sdk(plainEp);

    // ------------------------------------------------------------------ 1
    // Default: KMS object, cacheKms=false. Nothing may reach the tier.
    conn.sync().flushall();
    try (var sdkc = new S3SdkObjectClient(kmsS3, false)) {
      var c = new FlintObjectClient(sdkc, conn.async(), FlintObjectClient.DEFAULT_CHUNK_BYTES, 50, 2,
          false, kmsS3, false);
      byte[] got = Suite.read(c, KEY, 0, LEN);
      check(Suite.genOf(KEY, 0, got) == 0, "KMS object still reads CORRECTLY (bypass is not a failure)");
      long keys = conn.sync().keys("*".getBytes()).size();
      check(keys == 0, "and NOTHING reached the tier -- no chunks, no metadata (" + keys + " keys)");
      check(c.kmsBypassed.get() > 0,
          "armed: it was the KMS path that bypassed (" + c.kmsBypassed.get() + "), not some other bypass");
      check(c.kmsUndetectable.get() == 0,
          "armed: detection actually worked -- 0 undetectable objects");
    }

    // ------------------------------------------------------------------ 2
    // The control that carries check 1. A client that cached nothing at all
    // would pass every assertion above, so the SAME configuration must cache a
    // non-KMS object from a non-KMS origin.
    conn.sync().flushall();
    try (var sdkc = new S3SdkObjectClient(plainS3, false)) {
      var c = new FlintObjectClient(sdkc, conn.async(), FlintObjectClient.DEFAULT_CHUNK_BYTES, 50, 2,
          false, plainS3, false);
      byte[] got = Suite.read(c, KEY, 0, LEN);
      check(Suite.genOf(KEY, 0, got) == 0, "control: a PLAIN object reads correctly");
      long keys = conn.sync().keys("c2/*".getBytes()).size();
      check(keys > 0,
          "control: and the same client DOES cache it (" + keys + " chunks) -- so check 1 "
          + "measured the KMS rule, not a broken cache");
      check(c.kmsBypassed.get() == 0, "control: and nothing was KMS-bypassed");
    }

    // ------------------------------------------------------------------ 3
    // Opt-in. The customer has decided; the same KMS object must now cache.
    conn.sync().flushall();
    try (var sdkc = new S3SdkObjectClient(kmsS3, false)) {
      var c = new FlintObjectClient(sdkc, conn.async(), FlintObjectClient.DEFAULT_CHUNK_BYTES, 50, 2,
          false, kmsS3, true);
      byte[] got = Suite.read(c, KEY, 0, LEN);
      check(Suite.genOf(KEY, 0, got) == 0, "opt-in: the KMS object reads correctly");
      long keys = conn.sync().keys("c2/*".getBytes()).size();
      check(keys > 0, "OPT-IN WORKS: flint.cache.sse-kms=true caches the KMS object ("
          + keys + " chunks)");
      check(c.kmsBypassed.get() == 0, "armed: and nothing was bypassed once opted in");

      // and it is genuinely the cache, not a coincidence
      int before = Suite.gets();
      byte[] warm = Suite.read(c, KEY, 0, LEN);
      check(Arrays.equals(got, warm) && Suite.gets() == before,
          "opt-in: the second read is served entirely from the tier (0 origin GETs)");
    }

    // ------------------------------------------------------------------ 4
    // Detection unavailable must not look like detection succeeding. Built
    // without an SDK client, the client cannot check, and says so.
    conn.sync().flushall();
    try (var sdkc = new S3SdkObjectClient(kmsS3, false)) {
      var c = new FlintObjectClient(sdkc, conn.async(), FlintObjectClient.DEFAULT_CHUNK_BYTES, 50, 2);
      Suite.read(c, KEY, 0, LEN);
      check(c.kmsUndetectable.get() > 0,
          "no SDK client: the client reports it COULD NOT check (" + c.kmsUndetectable.get()
          + ") rather than reporting all-clear");
    }

    conn.close(); rc.shutdown(); kmsS3.close(); plainS3.close();
    System.out.println(ok ? "SSE-KMS SUITE PASSED" : "SSE-KMS SUITE FAILED");
    System.exit(ok ? 0 : 1);
  }
}
